# Data Quality & Cleaning

This notebook focuses on validating and improving the quality of the FinSight source data before further analysis and modeling.

## Objectives

- identify duplicate records and validate key uniqueness,
- analyze missing values,
- correct inconsistent data types,
- validate relationships between datasets,
- detect invalid or anomalous values,
- build reusable data validation checks,
- generate a data quality summary,
- save cleaned datasets to `data/processed/`.

> Raw data in `data/raw/` remains unchanged throughout the process.

In [2]:
import pandas as pd
import numpy as np
import json

print("FinSight environment ready")

FinSight environment ready


## Load Source Data

In [3]:
users = pd.read_csv("../data/raw/users_data.csv")
transactions = pd.read_csv("../data/raw/transactions_data.csv")
cards = pd.read_csv("../data/raw/cards_data.csv")
with open("../data/raw/mcc_codes.json", "r") as file:
    mcc_dict = json.load(file)
codes = pd.DataFrame(
    mcc_dict.items(),
    columns=["mcc", "description"]
)
with open("../data/raw/train_fraud_labels.json", "r") as file:
    fraud_dict = json.load(file)
fraud_target = fraud_dict["target"]
fraud = pd.DataFrame(
    fraud_target.items(),
    columns=["transaction_id", "is_fraud"]
)

print("Loading raw files ready")

Loading raw files ready


## Duplicate Validation

### Users

In [4]:
users.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `users`.

In [5]:
users["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each user has a unique identifier.

### Cards

In [6]:
cards.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `cards`.

#### Primary Key Duplicate Check

Check whether `id` contains duplicate values, as each card should have a unique identifier.

In [7]:
cards["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each card has a unique identifier.

### Transactions

In [8]:
transactions.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `transactions`.

In [9]:
transactions["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each transaction has a unique identifier.

#### MCC Codes

In [10]:
codes.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `codes`.

In [11]:
codes["mcc"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `mcc` values were found. Each code has a unique identifier.

#### Fraud Labels

In [12]:
fraud.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `fraud`.

In [13]:
fraud["transaction_id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `transaction_id` values were found. Each fraud has a unique identifier.

## Missing Values Validation

#### Users

In [14]:
users.isna().sum()

id                   0
current_age          0
retirement_age       0
birth_year           0
birth_month          0
gender               0
address              0
latitude             0
longitude            0
per_capita_income    0
yearly_income        0
total_debt           0
credit_score         0
num_credit_cards     0
dtype: int64

**Result:** No missing values were found in `users`.

#### Cards

In [15]:
cards.isna().sum()

id                       0
client_id                0
card_brand               0
card_type                0
card_number              0
expires                  0
cvv                      0
has_chip                 0
num_cards_issued         0
credit_limit             0
acct_open_date           0
year_pin_last_changed    0
card_on_dark_web         0
dtype: int64

**Result:** No missing values were found in `cards`.

#### Transactions

In [16]:
transactions.isna().sum()

id                       0
date                     0
client_id                0
card_id                  0
amount                   0
use_chip                 0
merchant_id              0
merchant_city            0
merchant_state     1563700
zip                1652706
mcc                      0
errors            13094522
dtype: int64

#### Missing Values Percentage

Calculate the percentage of missing values in each column of `transactions`.

In [17]:
percentage = round(((transactions.isna().sum())/len(transactions)) * 100, 2)
percentage

id                 0.00
date               0.00
client_id          0.00
card_id            0.00
amount             0.00
use_chip           0.00
merchant_id        0.00
merchant_city      0.00
merchant_state    11.75
zip               12.42
mcc                0.00
errors            98.41
dtype: float64

#### Transaction Errors

Inspect the values present in `errors` to determine whether missing values represent transactions without an error.

In [18]:
transactions["errors"].value_counts()

errors
Insufficient Balance                                   130902
Bad PIN                                                 32119
Technical Glitch                                        26271
Bad Card Number                                          7767
Bad Expiration                                           6161
Bad CVV                                                  6106
Bad Zipcode                                              1126
Bad PIN,Insufficient Balance                              293
Insufficient Balance,Technical Glitch                     243
Bad Card Number,Insufficient Balance                       71
Bad PIN,Technical Glitch                                   70
Bad CVV,Insufficient Balance                               57
Bad Expiration,Insufficient Balance                        47
Bad Card Number,Bad CVV                                    38
Bad Card Number,Bad Expiration                             33
Bad Expiration,Bad CVV                                     32
B

#### Transaction Errors — Missing Values

Check whether missing values in `errors` represent transactions with no recorded error by including missing values in the frequency count.

In [19]:
transactions["errors"].value_counts(dropna = False)

errors
NaN                                                    13094522
Insufficient Balance                                     130902
Bad PIN                                                   32119
Technical Glitch                                          26271
Bad Card Number                                            7767
Bad Expiration                                             6161
Bad CVV                                                    6106
Bad Zipcode                                                1126
Bad PIN,Insufficient Balance                                293
Insufficient Balance,Technical Glitch                       243
Bad Card Number,Insufficient Balance                         71
Bad PIN,Technical Glitch                                     70
Bad CVV,Insufficient Balance                                 57
Bad Expiration,Insufficient Balance                          47
Bad Card Number,Bad CVV                                      38
Bad Card Number,Bad Expiration   

**Result:** The `errors` column contains 98.41% missing values. These values represent transactions with no recorded error rather than a data quality issue. Non-null values contain specific transaction error types.

#### Missing `merchant_state` Investigation

Inspect transactions with missing `merchant_state` values to determine whether the missingness is expected or indicates a data quality issue.

In [20]:
transactions[transactions["merchant_state"].isna()].head(5)

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
7,7475335,2010-01-01 00:14:00,1684,2140,$26.46,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
8,7475336,2010-01-01 00:21:00,335,5131,$261.58,Online Transaction,50292,ONLINE,NaN,NaN,7801,NaN
18,7475346,2010-01-01 00:34:00,394,4717,$26.04,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
24,7475353,2010-01-01 00:43:00,301,3742,$10.17,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
26,7475356,2010-01-01 00:45:00,566,3439,$16.86,Online Transaction,16798,ONLINE,NaN,NaN,4121,NaN


#### Missing `merchant_state` by Transaction Type

Check whether missing `merchant_state` values occur exclusively in online transactions.

In [21]:
transactions[
    transactions["merchant_state"].isna()
]["use_chip"].unique()

<StringArray>
['Online Transaction', 'Chip Transaction']
Length: 2, dtype: str

#### Chip Transactions with Missing `merchant_state`

Inspect chip transactions where `merchant_state` is missing to determine whether there is another pattern explaining the missing location data.

In [22]:
transactions[
    (transactions["merchant_state"].isna()) &
    (transactions["use_chip"] == "Chip Transaction")
].head()

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
6573529,15473515,2015-01-01 12:19:00,944,2828,$-134.00,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN
6573565,15473561,2015-01-01 12:27:00,944,2828,$358.94,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN
6573594,15473593,2015-01-01 12:32:00,1117,1162,$117.43,Chip Transaction,52073,ONLINE,NaN,NaN,4722,NaN
6573677,15473695,2015-01-01 12:51:00,1117,1162,$-147.00,Chip Transaction,52073,ONLINE,NaN,NaN,4722,NaN
6574120,15474225,2015-01-01 14:24:00,972,6052,$233.77,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN


#### Missing `merchant_state` by Merchant Location

Check whether transactions with missing `merchant_state` are associated with `merchant_city = ONLINE`.

In [23]:
transactions[
    transactions["merchant_state"].isna()
]["merchant_city"].value_counts()

merchant_city
ONLINE    1563700
Name: count, dtype: int64

**Result:** All 1,563,700 transactions with missing `merchant_state` have `merchant_city = ONLINE`. Therefore, the missing values are expected for online transactions and do not indicate a data quality issue.

#### Missing `zip` Investigation

Inspect transactions with missing `zip` values to determine whether the missingness is expected or indicates a data quality issue.

In [24]:
transactions[
    transactions["zip"].isna()
].head()

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
7,7475335,2010-01-01 00:14:00,1684,2140,$26.46,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
8,7475336,2010-01-01 00:21:00,335,5131,$261.58,Online Transaction,50292,ONLINE,NaN,NaN,7801,NaN
18,7475346,2010-01-01 00:34:00,394,4717,$26.04,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
24,7475353,2010-01-01 00:43:00,301,3742,$10.17,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
26,7475356,2010-01-01 00:45:00,566,3439,$16.86,Online Transaction,16798,ONLINE,NaN,NaN,4121,NaN


#### Missing `zip` by Merchant Location

Check which merchant locations are associated with missing `zip` values.

In [25]:
transactions[
    transactions["zip"].isna()
]["merchant_city"].value_counts()

merchant_city
ONLINE            1563700
Cancun               9325
Rome                 7081
Mexico City          5102
Cabo San Lucas       4550
                   ...   
Addis Ababa             3
Libreville              3
Paramaribo              3
Nuku Alofa              2
Baku                    2
Name: count, Length: 169, dtype: int64

#### Missing `zip` by Merchant State

Inspect `merchant_state` values for transactions with missing `zip` to determine whether the additional missing ZIP codes are associated with international transactions.

In [26]:
transactions[
    transactions["zip"].isna()
]["merchant_state"].value_counts(dropna=False)

merchant_state
NaN               1563700
Mexico              27401
Canada              10647
Italy                7081
United Kingdom       4482
                   ...   
Ethiopia                3
Gabon                   3
Suriname                3
Tonga                   2
Azerbaijan              2
Name: count, Length: 148, dtype: int64

**Result:** Missing `zip` values are primarily associated with online and international transactions. Online transactions have no physical merchant location, while international transactions may not use the ZIP code format represented by this field. Therefore, these missing values appear to be structurally expected rather than a data quality issue.

#### Mcc codes

In [27]:
codes.isna().sum()

mcc            0
description    0
dtype: int64

**Result:** No missing values were found in `codes`.

#### Fraud Labels

In [28]:
fraud.isna().sum()

transaction_id    0
is_fraud          0
dtype: int64

**Result:** No missing values were found in `fraud`.

## Data Type Validation

Review column data types and convert fields where necessary to ensure dates, numeric values, and categorical data are represented correctly.

### Users

Inspect the current data types in `users`.

In [29]:
users.dtypes

id                     int64
current_age            int64
retirement_age         int64
birth_year             int64
birth_month            int64
gender                   str
address                  str
latitude             float64
longitude            float64
per_capita_income        str
yearly_income            str
total_debt               str
credit_score           int64
num_credit_cards       int64
dtype: object

#### Monetary Columns

Inspect monetary columns currently stored as strings before converting them to numeric data types.

In [30]:
users[["per_capita_income", "yearly_income", "total_debt"]].head()

,per_capita_income,yearly_income,total_debt
0,$29278,$59696,$127613
1,$37891,$77254,$191349
2,$22681,$33483,$196
3,$163145,$249925,$202328
4,$53797,$109687,$183855


In [31]:
users["per_capita_income"] = (
    users["per_capita_income"]
    .str.replace("$", "", regex=False)
)
users["per_capita_income"].head()


0     29278
1     37891
2     22681
3    163145
4     53797
Name: per_capita_income, dtype: str

In [32]:
users["per_capita_income"] = pd.to_numeric(
    users["per_capita_income"]
)
users["per_capita_income"].dtype

dtype('int64')

In [33]:
users["yearly_income"] = (
    users["yearly_income"]
    .str.replace("$", "", regex=False)
)
users["yearly_income"].head()

0     59696
1     77254
2     33483
3    249925
4    109687
Name: yearly_income, dtype: str

In [34]:
users["yearly_income"] = pd.to_numeric(
    users["yearly_income"]
)
users["yearly_income"].dtype

dtype('int64')

In [35]:
users["total_debt"] = (
    users["total_debt"]
    .str.replace("$", "", regex=False)
)
users["total_debt"].head()

0    127613
1    191349
2       196
3    202328
4    183855
Name: total_debt, dtype: str

In [36]:
users["total_debt"] = pd.to_numeric(
    users["total_debt"]
)
users["total_debt"].dtype

dtype('int64')

**Result:** The monetary columns `per_capita_income`, `yearly_income`, and `total_debt` were converted from strings containing currency symbols to numeric (`int64`) values. The remaining columns already use appropriate data types.

### Users

Inspect the current data types in `users`.

In [37]:
cards.dtypes

id                       int64
client_id                int64
card_brand                 str
card_type                  str
card_number              int64
expires                    str
cvv                      int64
has_chip                   str
num_cards_issued         int64
credit_limit               str
acct_open_date             str
year_pin_last_changed    int64
card_on_dark_web           str
dtype: object

#### Columns Requiring Inspection

Inspect columns that may require data type conversion before applying any transformations.

In [38]:
cards[
    ["expires", "has_chip", "credit_limit", "acct_open_date", "card_on_dark_web"]
].head()

,expires,has_chip,credit_limit,acct_open_date,card_on_dark_web
0,12/2022,YES,$24295,09/2002,No
1,12/2020,YES,$21968,04/2014,No
2,02/2024,YES,$46414,07/2003,No
3,08/2024,NO,$12400,01/2003,No
4,03/2009,YES,$28,09/2008,No


#### Convert `expires`

Convert the card expiration month and year from string format to datetime.

In [39]:
cards["expires"] = pd.to_datetime(
    cards["expires"],
    format="%m/%Y"
)
cards["expires"].head()

0   2022-12-01
1   2020-12-01
2   2024-02-01
3   2024-08-01
4   2009-03-01
Name: expires, dtype: datetime64[us]

#### Convert `has_chip`

Convert `has_chip` from `YES` / `NO` string values to boolean values for clearer and more efficient analysis. 

In [40]:
cards["has_chip"].value_counts(dropna=False)

has_chip
YES    5500
NO      646
Name: count, dtype: int64

In [41]:
cards["has_chip"] = cards["has_chip"].map({
    "YES": True,
    "NO": False                                      
})
cards["has_chip"].head()

0     True
1     True
2     True
3    False
4     True
Name: has_chip, dtype: bool

In [42]:
cards["has_chip"].dtype

dtype('bool')

#### Convert `credit_limit`

Remove the currency symbol from `credit_limit` and convert the column from string to a numeric data type.

In [43]:
cards["credit_limit"] = (
    cards["credit_limit"]
    .str.replace("$", "", regex=False)
)
cards["credit_limit"].head()

0    24295
1    21968
2    46414
3    12400
4       28
Name: credit_limit, dtype: str

In [44]:
cards["credit_limit"] = pd.to_numeric(
    cards["credit_limit"]
)
cards["credit_limit"].dtype

dtype('int64')

#### Convert `acct_open_date`

Convert `acct_open_date` from month/year string format to datetime.

In [45]:
cards["acct_open_date"] = pd.to_datetime(
    cards["acct_open_date"],
    format="%m/%Y"
)
cards["acct_open_date"].head()

0   2002-09-01
1   2014-04-01
2   2003-07-01
3   2003-01-01
4   2008-09-01
Name: acct_open_date, dtype: datetime64[us]

#### Convert `card_on_dark_web`

Convert `card_on_dark_web` from `Yes` / `No` string values to boolean values for clearer analysis.

In [46]:
cards["card_on_dark_web"].value_counts(dropna=False)

card_on_dark_web
No    6146
Name: count, dtype: int64

In [47]:
cards["card_on_dark_web"] = cards["card_on_dark_web"].map({
    "Yes": True,
    "No": False
})

cards["card_on_dark_web"].dtype

dtype('bool')

### Transactions

Inspect the current data types in `transactions` and identify columns that require conversion.

In [48]:
transactions.dtypes

id                  int64
date                  str
client_id           int64
card_id             int64
amount                str
use_chip              str
merchant_id         int64
merchant_city         str
merchant_state        str
zip               float64
mcc                 int64
errors                str
dtype: object

In [49]:
transactions[["date", "amount"]].head()

,date,amount
0,2010-01-01 00:01:00,$-77.00
1,2010-01-01 00:02:00,$14.57
2,2010-01-01 00:02:00,$80.00
3,2010-01-01 00:05:00,$200.00
4,2010-01-01 00:06:00,$46.41


#### Convert `date`

Convert the transaction timestamp from string format to datetime for time-based analysis.

In [50]:
transactions["date"] = pd.to_datetime(
    transactions["date"],
    format="%Y-%m-%d %H:%M:%S"
)

transactions["date"].dtype

dtype('<M8[us]')

#### Convert `amount`

Remove the currency symbol from `amount` and convert the column from string to a numeric data type while preserving decimal values.

In [51]:
transactions["amount"] = (
    transactions["amount"]
    .str.replace("$", "", regex = False)
)
transactions["amount"].head()

0    -77.00
1     14.57
2     80.00
3    200.00
4     46.41
Name: amount, dtype: str

In [52]:
transactions["amount"] = pd.to_numeric(
    transactions["amount"]
)
transactions["amount"].dtype

dtype('float64')

### MCC Codes

Inspect the current data types in `codes` and identify whether any columns require conversion.

In [53]:
codes.dtypes

mcc            str
description    str
dtype: object

In [54]:
codes["mcc"] = pd.to_numeric(
    codes["mcc"]
)
codes["mcc"].dtype

dtype('int64')

### Fraud Labels

Inspect the current data types in `fraud` and identify whether any columns require conversion.

In [55]:
fraud.dtypes

transaction_id    str
is_fraud          str
dtype: object

In [56]:
fraud["transaction_id"] = pd.to_numeric(
    fraud["transaction_id"]
)
fraud["transaction_id"].dtype

dtype('int64')

In [57]:
fraud["is_fraud"].value_counts()

is_fraud
No     8901631
Yes      13332
Name: count, dtype: int64

In [58]:
fraud["is_fraud"] = fraud["is_fraud"].map({
    "Yes": True,
    "No": False                                      
})
fraud["is_fraud"].head()

0    False
1    False
2    False
3    False
4    False
Name: is_fraud, dtype: bool

## Relationship Integrity Validation

Validate key relationships between datasets to identify transactions referencing non-existent users, cards, or MCC codes.

### Transactions → Users

Check whether every `client_id` in `transactions` corresponds to an existing `id` in `users`.

_merge
both          13305915
left_only            0
right_only           0
Name: count, dtype: int64